### ISSUE TITLE
Found a way to to use any markup in howm

### ISSUE THREAD

**Issue by artsi0m** [kaorahi/howm#7]
![image](https://github.com/kaorahi/howm/assets/87656471/df9d76df-3b84-4c82-9d3e-9ebeae399db7)

I created separate repository to tinker with howm ability to use any  possible markup.
https://github.com/artsi0m/howm-use-any-markup

What I want to reach is set of command like `howm-org-mode` that will change major mode, apply howm as minor mode and then add local variable for automatic detection of major mode.

It would be nice to merge such functionality in the howm itself one day, but I can cope well with preserving it as a separate package in the future.

I think there would be some issues with using this markups, but I will report them in this thread and will try to fix.

**Comment by artsi0m** [comment]
Stable version of a function for inserting prop-line in howm from repo mentioned before.
For now it is interactive only.

```elisp
    (defun howm-insert-prop-line ()
  "Activate major mode and modify the file so that this mode is activated
  automatically the next time it is opened"
      (interactive)
      (howm-mode)
      (let*
	  ((modes (mapcar #'cdr auto-mode-alist))
	   (mode-name (completing-read "Choose major mode: " modes))
	   (mode (intern-soft mode-name)))
	(unless (or (null mode)
	    (eq mode major-mode))
	  (funcall mode)
	  (howm-mode)
	  (add-file-local-variable-prop-line
	   'mode (intern (string-trim-right mode-name "-mode\\'"))))))
```

**Comment by artsi0m** [comment]
` sed "s:^\x2A := :" -i */*/*.org` — sed BRE to change only first level of org headings `*` to howm heading `=`

BTW: In asciidoc heading would start with `=` too, and continue with `==` similarly to org-mode.

**Comment by artsi0m** [comment]
Memory consuming script to add prop-line specifying org major mode to every file in howm that exists.
 
```python
# Script to apply local var to all org files
from pathlib import Path
from os import rename

files = list(Path(".").rglob("*.org"))

for el in files:
    with open(el,'r', encoding="utf-8") as contents:
          save = contents.read()
    with open(el,'w', encoding="utf-8") as contents:
          contents.write("# -*- mode: org -*-\n")
          contents.write(save)
```

**Comment by artsi0m** [comment]
``` lisp
      (defun howm-insert-prop-line ()
  "Activate major mode and modify the file so that this mode is activated
  automatically the next time it is opened"
      (interactive)
      (howm-mode)
      (let*
	  ((modes (mapcar #'cdr auto-mode-alist))
	   (mode-name (completing-read "Choose major mode: " modes))
	   (mode (intern-soft mode-name)))
	(unless (or (null mode)
	    (eq mode major-mode))
	  (funcall mode)
	  (howm-mode)
	  (add-file-local-variable-prop-line
	   'mode (intern (string-trim-right mode-name "-mode\\'"))))))

```

That's my `howm-insert-prop-line` function for now.

What do I have to do to get it merged ?

**Comment by kaorahi** [comment]
Thanks for the proposal. This feature seems useful for broader use cases, even outside of howm. Is there a reason to make it specific to howm? (Or am I misunderstanding something?)

**Comment by artsi0m** [comment]
> Is there a reason to make it specific to howm?

Yes, it is. It calls `(howm-mode)` two times, so when major mode will be enabled it also enables howm mode, without those calls to `(howm-mode)` it would just enable chosen major mode.

I wrote function this way, so it would enable both chosen major mode and howm minor mode.

In my opinion it is specific to howm workflow, because only in howm you may want to assign some major mode to a file without changing it extension. But of course I can create a general version with broader use case.

**Comment by artsi0m** [comment]
>But of course I can greate general version with broader use case.

That would be true when I will find how I can preserve enabled minor modes across changing of major mode.

**Comment by kaorahi** [comment]
generic version based on your code?

```emacs-lisp
(defun my-persist-major-mode (mode)
  "Activate major mode MODE and modify the file so that this mode is activated
  automatically the next time it is opened"
  (interactive (let ((modes
                      (seq-filter (lambda (mode)
                                    (and (symbolp mode) (commandp mode)))
                                  (delete-dups (mapcar #'cdr auto-mode-alist)))))
                 ;; Avoid using mode-name here, as it is a built-in variable.
                 (list
                  (intern-soft (completing-read "Choose major mode: " modes)))))
  (when (or (null mode) (eq mode major-mode))
    (error "Already in %s." major-mode))
  (let* ((candidate-minor-modes minor-mode-list)
         ;; or (candidate-minor-modes '(howm-mode foo-mode bar-mode ...))
         (minor-modes-to-reapply
          (seq-filter (lambda (m) (and (boundp m) (symbol-value m)))
                      candidate-minor-modes)))
    (funcall mode)
    (mapc (lambda (minor-mode)
            (unless (and (boundp minor-mode) (symbol-value minor-mode))
              (when (fboundp minor-mode)
                (message "Reapply %s." minor-mode)
                (funcall minor-mode))))
          minor-modes-to-reapply)
    (delete-file-local-variable-prop-line 'mode)
    (add-file-local-variable-prop-line
     'mode (intern (string-trim-right (symbol-name mode) "-mode\\'")))))

;; (setq enable-local-variables t)
```

**Comment by artsi0m** [comment]
How to become such experienced programmer as you ?

Thanks for the generic version of code.

**Comment by kaorahi** [comment]
I'll close this issue now since I've added a link to it in the Emacs Wiki. Feel free to reopen it if you still have any problems.

**Comment by artsi0m** [comment]
> I've added a link to it in the Emacs Wiki ("More Tips").

Thanks.

> Feel free to reopen it if you still have any problems.

Everything is alright by now. But github says that I don't have permission to reopen this issue.

**Comment by artsi0m** [comment]
It would be nice if you reopen this issue. I wrote an README with annotated screenshot and now howm-use-any-markup package could be installed using vc-use-package.

**Comment by kaorahi** [comment]
Thank you for the notification. I'll reopen this now.

**Comment by reyman** [comment]
> It would be nice if you reopen this issue. I wrote an README with annotated screenshot and now howm-use-any-markup package could be installed using vc-use-package.

Hi !

I'm interested to use both markdown and org-mode for different repositories of notes (multiple project, different team), is it possible to get mixed mode with your plugin @artsi0m ? 

If this is the case, how i declare the  `use-package` part, both `require 'howm-org` and `require 'howm-markdown` are declared  ? 

More generally it could be nice to set mode by folder, for example :

```lisp
(setq howm-search-path '(list ("/home/reyman/.../Notes/OrganicMatter" "org-mode) 
("/home/reyman/.../Notes/Ecotone" "markdown-mode") ) )
```

**Comment by artsi0m** [comment]
> I'm interested to use both markdown and org-mode for different repositories of notes (multiple project, different team), is it
> possible to get mixed mode with your plugin @artsi0m ?

It is possible to mix theme even inside the same repository.

> If this is the case, how i declare the use-package part, both require 'howm-org and require 'howm-markdown are declared ? 

You do neither. I mean that I do neither in my own config file.

>how i declare the use-package part

I believe it would look something like this:

```

(use-package howm
  :custom
  (howm-search-other-dir t))

(use-package howm-use-any-markup
  :vc (:url "https://github.com/artsi0m/howm-use-any-markup")
  :ensure t)

;; Here is exact snippet, although not from my config, but from evaluated note itself
(push "~/Projects/PlaysDev.d/PlaysDev-tasks/task-15" howm-search-path)
(delete-dups howm-search-path)

```

After that:
1. Create a note with `M-x` `howm-create` or open file in another folder
2. `M-x` `howm-insert-prop-line` or `M-x` `howm-insert-local-var`
3. Completion menu would appear, suggesting list of the major
modes. Choose major mode for your note.

After that every time you would open that note emacs would select
dedicated major mode that you choose before.

Sorry for the late reply. I hope that now it's clear.

>More generally it could be nice to set mode by folder

I don't know how to do this right now. Although I think that it is achievable my package was written for the other purpose of selecting desired markup per note.

You probably could advice `howm-create` to call `howm-insert-prop-line` or `howm-insert-local-var` non-interactively with desired mode based of the folder. Yeah, I am sorry, I would think about it more tomorrow.
